Network/AMQP/Internal.hs

```diff
@@ -453,7 +453,8 @@ data Channel = Channel {
                     chanActive :: Lock, -- used for flow-control. if lock is closed, no content methods will be sent
                     chanClosed :: MVar (Maybe String),
                     consumers :: MVar (M.Map Text ((Message, Envelope) -> IO ())), -- who is consumer of a queue? (consumerTag => callback)
-                    returnListeners :: MVar ([(Message, PublishError) -> IO ()])
+                    returnListeners :: MVar ([(Message, PublishError) -> IO ()]),
+                    returnHandlers :: MVar [CE.SomeException -> IO ()]
```

Comment:
The idea here is that in the callback you are passed an exception, so you can actually know what caused the thread to die.